src/mpid/ch4/netmod/ofi/ofi_impl.h

```diff
@@ -42,14 +42,16 @@ int MPIDI_OFI_progress(int vci, int blocking);
 /* vni mapping */
 /* NOTE: concerned by the modulo? If we restrict num_vnis to power of 2,
  * we may get away with bit mask */
-MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_get_vni_src(MPIR_Comm * comm_ptr, int rank, int tag)
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_get_vni_src(MPIR_Comm * comm_ptr,
+                                                   int src_rank, int dst_rank, int tag)
 {
-    return MPIDI_vci_get_src(comm_ptr, rank, tag) % MPIDI_OFI_global.num_vnis;
+    return MPIDI_vci_get_src(comm_ptr, src_rank, dst_rank, tag) % MPIDI_OFI_global.num_vnis;
```

Comment:
@hzhou In our implementation, we have new other differences. Here's the interface corresponding to `MPIDI_vci_get_src`: `MPIDI_get_sender_vci(MPIR_Comm * comm, MPIR_Context_id_t ctxid_in_effect, int sender_rank, int receiver_rank, int tag)`. Here `ctxid_in_effect`  is used for hashing purpose and can be either `comm->context_id` or `comm->recvcontext_id`, depending on situation that is available from where this routine is called.  We still need comm object to access other things including the hints. 

Similar difference for `receiver` hash.  

I think it would be worth addressing those difference in the PR.

Reply:
How about an enum *context_type*? You don't really need repeat the context_id, since it is accessible in the comm pointer; you just need a flag or type tell you which one to use. In addition, a general enum type can also mark whether the message is collective, pt2pt, shmem or netmod or active message, then we can have more specialized and fancier hashing possibilities.

Reply:
I think that should work as both of the `context_id` values are present inside the comm.